.github/workflows/docker.yml

```diff
@@ -60,380 +62,86 @@ concurrency:
-        x86_64-minimal,
-        x86_64-scipy,
-        x86_64-r,
-        x86_64-julia,
-        x86_64-tensorflow,
-        x86_64-pytorch,
-        x86_64-datascience,
-        x86_64-pyspark,
-        x86_64-all-spark,
-      ]
-    if: ${{ !contains(github.event.pull_request.title, '[FAST_BUILD]') }}
+    if: github.repository_owner == 'jupyter'
 
-  x86_64-images-tag-push-fast:
-    uses: ./.github/workflows/docker-tag-push.yml
+  build-x86_64:
+    uses: ./.github/workflows/docker-build-test-upload-all.yml
     with:
       platform: x86_64
-      image: ${{ matrix.image }}
+      runs-on-fast: ubuntu-latest
+      runs-on-slow: ubuntu-latest
     secrets:
       REGISTRY_USERNAME: ${{ secrets.QUAY_USERNAME }}
       REGISTRY_TOKEN: ${{ secrets.QUAY_ROBOT_TOKEN }}
-    strategy:
-      matrix:
-        image: [docker-stacks-foundation, base-notebook]
-    needs: [x86_64-foundation, x86_64-base]
-    if: contains(github.event.pull_request.title, '[FAST_BUILD]')
 
   merge-tags:
     uses: ./.github/workflows/docker-merge-tags.yml
     with:
-      image: ${{ matrix.image }}
+      image: ${{ matrix.image-variant.image }}
+      variant: ${{ matrix.image-variant.variant }}
     secrets:
       REGISTRY_USERNAME: ${{ secrets.QUAY_USERNAME }}
       REGISTRY_TOKEN: ${{ secrets.QUAY_ROBOT_TOKEN }}
     strategy:
       matrix:
-        image:
+        image-variant:
           [
-            docker-stacks-foundation,
-            base-notebook,
-            minimal-notebook,
-            scipy-notebook,
-            r-notebook,
-            julia-notebook,
-            tensorflow-notebook,
-            pytorch-notebook,
-            datascience-notebook,
-            pyspark-notebook,
-            all-spark-notebook,
+            { image: docker-stacks-foundation, variant: default },
+            { image: base-notebook, variant: default },
+            { image: minimal-notebook, variant: default },
+            { image: scipy-notebook, variant: default },
+            { image: r-notebook, variant: default },
+            { image: julia-notebook, variant: default },
+            { image: tensorflow-notebook, variant: default },
+            { image: pytorch-notebook, variant: default },
+            { image: pytorch-notebook, variant: cuda },
+            { image: pytorch-notebook, variant: cuda11 },
+            { image: datascience-notebook, variant: default },
+            { image: pyspark-notebook, variant: default },
+            { image: all-spark-notebook, variant: default },
           ]
-    needs: [aarch64-images-tag-push, x86_64-images-tag-push]
-    if: github.repository_owner == 'jupyter' && !contains(github.event.pull_request.title, '[FAST_BUILD]')
+    needs: [build-aarch64, build-x86_64]
+    if: |
+      always() &&
+      needs.build-x86_64.result == 'success' &&
+      (needs.build-aarch64.result == 'success' || needs.build-aarch64.result == 'skipped') &&
+      !contains(github.event.pull_request.title, '[FAST_BUILD]')
```

Comment:
Could you please explain why this need to this complicated?

Reply:
This is technically not needed for this repo. Should I remove it?

I added this for my own image builds, because skipped aarch64 builds do not count for `needs: [build-aarch64, build-x86_64]`. If this is skipped it will only push tags with the `x86_64_` prefix. So it will not have tags like `:hub-4.0.2` and only `x86_64-hub-4.0.2`.

Reply:
I see. No, I would like my fork to conitnue to work properly (you can see that sometimes I run some things only for `jupyter` GitHub user).
And you might have noticed that your fork works as it should too 🙂

Reply:
Added this adjusted check to both the `merge-tags-fast` and `merge-tags` (without the refactored ´-all.yaml`).